File: src/pieces/piece.rs
use crate::pieces::*;
use crate::{array::check_position_is_on_board, game::Colour};
use phf::phf_map;
use std::{fmt::Display};

use crate::types::{Array2D, PieceClass, Position, Shape};

static PIECE_CODE_HASHMAP: phf::Map<&'static str, PieceClass> = phf_map! {
    "B" => PieceClass::Bishop,
    "BISHOP" => PieceClass::Bishop,
    "K" => PieceClass::King,
    "KING" => PieceClass::King,
    "KNIGHT" => PieceClass::Knight,
    "N" => PieceClass::Knight,
    "P" => PieceClass::Pawn,
    "PAWN" => PieceClass::Pawn,
    "Q" => PieceClass::Queen,
    "QUEEN" => PieceClass::Queen,
    "R" => PieceClass::Rook,
    "ROOK" => PieceClass::Rook,
    "b" => PieceClass::Bishop,
    "bishop" => PieceClass::Bishop,
    "k" => PieceClass::King,
    "king" => PieceClass::King,
    "knight" => PieceClass::Knight,
    "n" => PieceClass::Knight,
    "p" => PieceClass::Pawn,
    "pawn" => PieceClass::Pawn,
    "q" => PieceClass::Queen,
    "queen" => PieceClass::Queen,
    "r" => PieceClass::Rook,
    "rook" => PieceClass::Rook,
};
#[derive(Clone, Debug, Hash, PartialEq, Eq)]
pub struct Piece {
    pub(crate) position: Position,
    pub(crate) symbol: String,
    pub(crate) full_symbol: PieceClass,
    pub(crate) is_white: bool,
    pub(crate) colour: Colour,
    pub(crate) in_play: bool,
    has_moved: bool,
}

impl Piece {
    pub fn new(position: Position, symbol: String) -> Self {
        let symbol_is_upper = symbol.to_uppercase() == symbol;

        Piece {
            position,
            symbol: symbol.clone(),
            full_symbol: PIECE_CODE_HASHMAP[&symbol as &str],
            is_white: symbol_is_upper,
            colour: if symbol_is_upper {
                Colour::White
            } else {
                Colour::Black
            },
            in_play: true,
            has_moved: false,
        }
    }
}

impl Display for Piece {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.symbol)
    }
}

fn allowed_moves(
    piece: &Piece,
    board: &Array2D,
    pos: &Position,
    shape: &Shape,
    is_white: bool,
) -> Vec<Position> {
    let full_symbol = piece.full_symbol;
    match full_symbol {
        PieceClass::King => KingBehaviour::allowed_moves(piece, board, pos, shape, is_white),
        PieceClass::Queen => QueenBehaviour::allowed_moves(piece, board, pos, shape, is_white),
        PieceClass::Rook => RookBehaviour::allowed_moves(piece, board, pos, shape, is_white),
        PieceClass::Bishop => BishopBehaviour::allowed_moves(piece, board, pos, shape, is_white),
        PieceClass::Knight => KnightBehaviour::allowed_moves(piece, board, pos, shape, is_white),
        PieceClass::Pawn => PawnBehaviour::allowed_moves(piece, board, pos, shape, is_white),
    }
}

pub(crate) fn get_allowed_moves(piece: &Piece, board: &Array2D, shape: &Shape) -> Vec<Position> {
    let new_positions = allowed_moves(piece, board, &piece.position, shape, piece.is_white);
    let new_positions = new_positions
        .into_iter()
        .filter(|pos| check_position_is_on_board(pos, shape))
        .collect();
    new_positions
}

fn allowed_takes(
    piece: &Piece,
    board: &Array2D,
    pos: &Position,
    shape: &Shape,
    is_white: bool,
) -> Vec<Position> {
    let full_symbol = piece.full_symbol;
    match full_symbol {
        PieceClass::King => KingBehaviour::allowed_takes(piece, board, pos, shape, is_white),
        PieceClass::Queen => QueenBehaviour::allowed_takes(piece, board, pos, shape, is_white),
        PieceClass::Rook => RookBehaviour::allowed_takes(piece, board, pos, shape, is_white),
        PieceClass::Bishop => BishopBehaviour::allowed_takes(piece, board, pos, shape, is_white),
        PieceClass::Knight => KnightBehaviour::allowed_takes(piece, board, pos, shape, is_white),
        PieceClass::Pawn => PawnBehaviour::allowed_takes(piece, board, pos, shape, is_white),
    }
}

pub(crate) fn get_allowed_takes(piece: &Piece, board: &Array2D, shape: &Shape) -> Vec<Position> {
    let new_positions = allowed_takes(piece, board, &piece.position, shape, piece.is_white);
    let new_positions = new_positions
        .into_iter()
        .filter(|pos| check_position_is_on_board(pos, shape))
        .collect();
    new_positions
}

pub(crate) fn get_legal_moves(piece: &Piece, board: &Array2D, shape: &Shape) -> Vec<Position> {
    let mut allowed_moves: Vec<Position> = get_allowed_moves(piece, board, shape);
    let mut allowed_takes: Vec<Position> = get_allowed_takes(piece, board, shape);
    allowed_moves.append(&mut allowed_takes);
    allowed_moves.sort_unstable();
    allowed_moves.dedup();
    allowed_moves
}

File: src/pieces/Piece.ts
const KingBehaviour = require("KingBehaviour");

enum PieceCode {
    KING,
    QUEEN,
    ROOK,
    BISHOP,
    KNIGHT,
    PAWN
}


interface Piece {
	position: Position;
	symbol?: string;
	full_symbol?: PieceCode;
	is_white: boolean;
	colour?: Colour;
	in_play?: boolean;
	has_moved?: boolean;
    is_empty: boolean
}

const PIECE_CODE_OBJECT = {
    'BISHOP': PieceCode.BISHOP,
    'bishop': PieceCode.BISHOP,
    'B': PieceCode.BISHOP,
    'b': PieceCode.BISHOP,
    'ROOK': PieceCode.ROOK,
    'rook': PieceCode.ROOK,
    'R': PieceCode.ROOK,
    'r': PieceCode.ROOK,
    'KING': PieceCode.KING,
    'king': PieceCode.KING,
    'K': PieceCode.KING,
    'k': PieceCode.KING,
    'PAWN': PieceCode.PAWN,
    'pawn': PieceCode.PAWN,
    'P': PieceCode.PAWN,
    'p': PieceCode.PAWN,
    'QUEEN': PieceCode.QUEEN,
    'queen': PieceCode.QUEEN,
    'Q': PieceCode.QUEEN,
    'q': PieceCode.QUEEN,
    'KNIGHT': PieceCode.KNIGHT,
    'knight': PieceCode.KNIGHT,
    'N': PieceCode.KNIGHT,
    'n': PieceCode.KNIGHT
}
const PIECE_CODE_DICT: Map<string, PieceCode> = new Map(Object.entries(PIECE_CODE_OBJECT));

function make_piece(pos: Position, symbol: string): Piece {
    const is_white = symbol.toUpperCase() === symbol
    return {
        position: pos,
        symbol: symbol,
        full_symbol: PIECE_CODE_DICT.get(symbol),
        is_white: is_white,
        colour: is_white ? Colour.White : Colour.Black,
        in_play: true,
        has_moved: false,
    } as Piece
}

function allowed_moves(piece: Piece, board: Array2D, pos: Position, is_white: boolean): Position[] {
    const full_symbol = piece.full_symbol
    let allowed_moves_array: Position[] = []
    if (full_symbol === PieceCode.KING) {
        allowed_moves_array = king_allowed_moves(piece, board, pos, is_white)
    } else if (full_symbol === PieceCode.QUEEN) {
        allowed_moves_array = queen_allowed_moves(piece, board, pos, is_white)
    } else if (full_symbol === PieceCode.ROOK) {
        allowed_moves_array = rook_allowed_moves(piece, board, pos, is_white)
    } else if (full_symbol === PieceCode.BISHOP) {
        allowed_moves_array = bishop_allowed_moves(piece, board, pos, is_white)
    } else if (full_symbol === PieceCode.KNIGHT) {
        allowed_moves_array = knight_allowed_moves(piece, board, pos, is_white)
    } else if (full_symbol === PieceCode.PAWN) {
        allowed_moves_array = pawn_allowed_moves(piece, board, pos, is_white)
    }
    return allowed_moves_array
}

function get_allowed_moves(piece: Piece, board: Array2D): Position[] {
    const pos = piece.position
    let new_positions = allowed_moves(piece, board, pos, piece.is_white)
    return new_positions.filter((p) => check_position_is_on_board(p, board.shape))
}

function allowed_takes(piece: Piece, board: Array2D, pos: Position, is_white: boolean): Position[] {
    const full_symbol = piece.full_symbol
    let allowed_takes_array: Position[] = []
    if (full_symbol === PieceCode.KING) {
        allowed_takes_array = king_allowed_takes(piece, board, pos, is_white)
    } else if (full_symbol === PieceCode.QUEEN) {
        allowed_takes_array = queen_allowed_takes(piece, board, pos, is_white)
    } else if (full_symbol === PieceCode.ROOK) {
        allowed_takes_array = rook_allowed_takes(piece, board, pos, is_white)
    } else if (full_symbol === PieceCode.BISHOP) {
        allowed_takes_array = bishop_allowed_takes(piece, board, pos, is_white)
    } else if (full_symbol === PieceCode.KNIGHT) {
        allowed_takes_array = knight_allowed_takes(piece, board, pos, is_white)
    } else if (full_symbol === PieceCode.PAWN) {
        allowed_takes_array = pawn_allowed_takes(piece, board, pos, is_white)
    }
    return allowed_takes_array
}

function get_allowed_takes(piece: Piece, board: Array2D): Position[] {
    const pos = piece.position
    let new_positions = allowed_takes(piece, board, pos, piece.is_white)
    return new_positions.filter((p) => check_position_is_on_board(p, board.shape))
}

function get_legal_moves(piece: Piece, board: Array2D): Position[] {
    const allowed_moves = get_allowed_moves(piece, board)
    const allowed_takes = get_allowed_takes(piece, board)
    const allowed: Position[] = [...allowed_moves, ...allowed_takes]
    return [...new Set(allowed)]
}

